universum/analyzers/utils.py

```diff
@@ -4,36 +4,91 @@
 import glob
 import subprocess
 
-from typing import Callable, List
+from typing import Any, Callable, List, Optional, Tuple
 from typing_extensions import TypedDict
 
+from universum.lib.ci_exception import CiException
+
 ReportData = TypedDict('ReportData', {'path': str, 'message': str, 'symbol': str, 'line': int})
 
 
-def report_parsed_outcome(cmd: List[str],
-                          parse_function: Callable[[str], List[ReportData]],
-                          out_file: str) -> int:
+class AnalyzerException(CiException):
+    def __init__(self, code: int = 2, message: Optional[str] = None):
+        self.code: int = code
+        self.message: Optional[str] = message
+
+
+def analyzer(parser: argparse.ArgumentParser):
+    def internal(func: Callable[[argparse.Namespace], List[ReportData]]) -> Callable[[], List[ReportData]]:
+        def wrapper() -> List[ReportData]:
+            add_result_file_argument(parser)
+            settings: argparse.Namespace = parser.parse_args()
+            issues: List[ReportData] = func(settings)
+            report_to_file(issues, settings.result_file)
+            return issues
+
+        return wrapper
+
+    return internal
+
+
+def sys_exit(func: Callable[[], Any]) -> Callable[[], None]:
```

Comment:
From reading the code it is not clear why 'sys_exit' decorator is separate from 'analyzer' decorator.
sys_exit is only applied to functions that also have analyzer applied.
I suggest to merge them into one.

Reply:
sys_exit is a generic wrapper not limited to analyzers - it has no reference to analyzer logic and can be used for any function/module

Reply:
That is ok.
But I don't understand, why 'analyzer' cannot use it internally without the need for clients to apply it explicitly.

Reply:
it can, but these functions have different purposes logically, and can exist independently.
I see no good reason to add an interdependency here.
E.g. sys_exit is needed only to call the analyzer directly via cmd line. If we add other options to call analyzers, there'd be no need for sys_exit there, but 'analyzer' function would still be needed to preserve the interface.

Reply:
I would say that we could remove the 'sys_exit' functionality from the 'analyzer' decorator as soon as it is not needed anymore. This is because with the current option (analyzer are scripts) the 'sys_exit' functionality appears to be another part of the interface.
Anyway, it is not strictly required to merge it.

Reply:
another reason why I prefer to keep these decorators separate is that this makes debugging easier - by removing sys_exit from the analyzer code I can easily see the root cause for generated exceptions, even during pytest run

Reply:
Please mention this in the comment in the source code